inference/analyze.py

```diff
@@ -0,0 +1,515 @@
+            self.tile_nnzs = self.tile_nnzs.cuda(device=device)
+
+            plot_tiles = False
+            if plot_tiles:
+                nnzs = np.diff(self.tile_nnzs)
+                plt.hist(nnzs, bins=np.arange(nnzs.min(), nnzs.max() + 1))
+                plt.show()
+
+        self.tensor_compress_interleaved()
+        self.second_order_data = self.second_order_data.cuda(device=device)
+        self.first_order_data = self.first_order_data.cuda(device=device)
+        self.x = torch.ones(self.n).cuda(device=device).half()
+        self.y = torch.zeros(self.m).cuda(device=device).half()
+
+    def analyse_sparsity(self, dense, name, base_path):
+        plt.axis('off')
+        d = max_pool2d(dense, 16)
+        plt.imshow(d)
+        # plt.tight_layout()
+        d = max_pool2d(dense, 16)
+        perc_tiles_filled = ((d.flatten() > 0).sum() / d.flatten().shape[0])
+        density = ((dense.flatten() > 0).sum() / dense.flatten().shape[0])
+        Round = lambda x, n: eval(
+            '"%.' + str(int(n)) + 'f" % ' + repr(int(x) + round(float('.' + str(float(x)).split('.')[1]), n)))
+        plt.savefig(
+            os.path.join(base_path,
+                         f'visualization/{name}_ratiofilled={Round(perc_tiles_filled, 2)}_density={Round(density, 2)}.png'),
+            bbox_inches='tight', transparent=True, pad_inches=0, dpi=400)
+
+    def run(self, num_runs, feature_flag):
+        runs = torch.empty(num_runs).cpu().float()
+
+        for i in range(num_runs):
+            self.y.zero_()
+            spqr_cuda.spqr_mul_timer(
+                self.m,
+                self.n,
+                self.args.bits,
+                self.args.beta1,
+                self.args.beta2,
+                self.first_order_data,
+                self.second_order_data,
+                self.values,
+                self.row_offsets,
+                self.col_ptr,
+                self.tile_row_count,
+                self.tile_count,
+                self.tile_row_offsets,
+                self.tile_col_ptr,
+                self.tile_nnzs,
+                self.tile_data,
+                self.x,
+                self.y,
+                self.nnz,
+                runs[i],
+                feature_flag)
+
+        time.sleep(1)
+        return runs, self.y.clone()
+
+    def compress_row(self, w, ws, wz):
+        pass
+
+    def tensor_compress_interleaved(self):
+        spqr_cuda.tensor_compress_interleaved(
+            self.m,
+            self.n,
+            self.args.bits,
+            self.W,
+            self.args.beta1,
+            self.args.beta2,
+            self.W_s,
+            self.W_z,
+            self.W_s_s,
+            self.W_s_z,
+            self.W_z_s,
+            self.W_z_z,
+            self.first_order_data,
+            self.second_order_data
+        )
+
+
+
+
+
+def updiv(x, y): return (x + y - 1) // y
+
+
+def densities():
+    fig = go.Figure()
+    device = torch.device('cuda:0')
```

Comment:
hard coded cuda:0

Reply:
I removed this script from the repo since it's just used to generate figures.